rend3-routine/src/skinning/mod.rs

```diff
@@ -0,0 +1,272 @@
+use rend3::{
+    format_sso,
+    graph::{DataHandle, RenderGraph},
+    managers::{MeshBuffers, SkeletonManager},
+    util::bind_merge::{BindGroupBuilder, BindGroupLayoutBuilder},
+};
+use wgpu::{BindGroupLayout, Buffer, BufferDescriptor, BufferUsages, ComputePipeline, Device};
+
+/// The per-skeleton data, as uploaded to the GPU compute shader.
+#[repr(C, align(16))]
+#[derive(Copy, Clone)]
+pub struct GpuSkinningInput {
+    /// See [rend3::managers::GpuVertexRanges].
+    pub mesh_range: [u32; 2],
+    /// See [rend3::managers::GpuVertexRanges].
+    pub skeleton_range: [u32; 2],
+    /// The index of this skeleton's first joint in the global joint matrix
+    /// buffer.
+    pub joint_idx: u32,
+}
+
+pub fn add_pre_skin_to_graph(graph: &mut RenderGraph, pre_skin_data: DataHandle<PreSkinningBuffers>) {
+    let mut builder = graph.add_node("pre-skinning");
+    let pre_skin_handle = builder.add_data_output(pre_skin_data);
+
+    builder.build(move |_pt, renderer, _encoder_or_pass, _temps, _ready, graph_data| {
+        let buffers = build_gpu_skinning_input_buffers(&renderer.device, graph_data.skeleton_manager);
+        graph_data.set_data::<PreSkinningBuffers>(pre_skin_handle, Some(buffers));
+    });
+}
+
+/// The two buffers uploaded to the GPU during pre-skinning.
+pub struct PreSkinningBuffers {
+    gpu_skinning_inputs: Buffer,
+    joint_matrices: Buffer,
+}
+
+fn build_gpu_skinning_input_buffers(device: &Device, skeleton_manager: &SkeletonManager) -> PreSkinningBuffers {
+    profiling::scope!("Building GPU Skinning Input Data");
+
+    let skinning_inputs_length = skeleton_manager.skeletons().len() * std::mem::size_of::<GpuSkinningInput>();
+    let gpu_skinning_inputs = device.create_buffer(&BufferDescriptor {
+        label: Some("skinning inputs"),
+        size: skinning_inputs_length as u64,
+        usage: BufferUsages::UNIFORM,
+        mapped_at_creation: true,
+    });
+
+    let joint_matrices_length: usize = skeleton_manager
+        .skeletons()
+        .map(|sk| sk.joint_matrices.len())
+        .sum::<usize>()
```

Comment:
We should statefully keep this number inside the skeleton manager to avoid needing a full pass over all skeletons.

Reply:
Took care of this. There's now a new `global_joint_count` field inside the `SkeletonManager` that's used instead. I believe I took care of updating that count at all the relevant places:

- In `SkeletonManager::fill`, incrementing by the necessary amount
- During `ready`, in `remove_all_dead`, where the amount is subtracted